### PR TITLE
[fix](storage) fix core for string predicate in storage layer

### PR DESCRIPTION
**Pull request by wangbo**
# Proposed changes

Currently, for string read from plain page, we hold ptr from page to avoid memcpy.
But if page is released, the ptr from page is invalid, core dump happens.
So for string from plain page, memcpy is neccessary.

## Problem Summary:

Describe the overview of changes.

## Checklist(Required)

1. Does it affect the original behavior: (Yes/No/I Don't know)
2. Has unit tests been added: (Yes/No/No Need)
3. Has document been added or modified: (Yes/No/No Need)
4. Does it need to update dependencies: (Yes/No)
5. Are there any changes that cannot be rolled back: (Yes/No)

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: be/src/vec/columns/predicate_column.h [hunk 1]
```diff
@@ -410,6 +429,9 @@ class PredicateColumnType final : public COWHelper<IColumn, PredicateColumnType<
 
 private:
     Container data;
+    // manages the memory for slice's data(For string type)
+    std::unique_ptr<MemPool> _pool;
+    bool _is_mem_pool_inited = false;;
```
Comment (yiguolei) [inline be/src/vec/columns/predicate_column.h:434]: just check _pool == nullptr?

File: be/src/vec/columns/predicate_column.h [hunk 2]
```diff
@@ -230,17 +230,36 @@ class PredicateColumnType final : public COWHelper<IColumn, PredicateColumnType<
...
             }
         }
     }
 
     void insert_default() override { data.push_back(T()); }
 
-    void clear() override { data.clear(); }
+    void clear() override { 
+        data.clear(); 
+        if (_is_mem_pool_inited) {
```
Comment (yiguolei) [inline be/src/vec/columns/predicate_column.h:259]: Will mempool release the memory when _pool is desconstructed?

File: be/src/vec/columns/predicate_column.h [hunk 3]
```diff
@@ -230,17 +230,36 @@ class PredicateColumnType final : public COWHelper<IColumn, PredicateColumnType<
...
+            for (size_t i = 0; i < num; i++) {
+                total_mem_size += len_array[i];
+            }
+            
+            char* destination = (char*)_pool->allocate(total_mem_size);
             for (size_t i = 0; i < num; i++) {
                 uint32_t len = len_array[i];
                 uint32_t start_offset = start_offset_array[i];
-                insert_string_value(data_array + start_offset, len);
+                memcpy(destination, data_array + start_offset, len);
```
Comment (jacktengg) [inline be/src/vec/columns/predicate_column.h:247]: Not sure how much performnce will be impacted by allocate and memcpy here.

File: be/src/vec/columns/predicate_column.h (hunk 2 above)
Comment (wangbo) [reply be/src/vec/columns/predicate_column.h:259]: This is just for memory reuse.
```
 mem_pool.h
   /// Makes all allocated chunks available for re-use, but doesn't delete any chunks.
    void clear();
```

File: be/src/vec/columns/predicate_column.h (hunk 1 above)
Comment (wangbo) [reply be/src/vec/columns/predicate_column.h:434]: 👌🏻

**Comment by morningman** [comment]
It seems that we are likely to refer to the memory in the page elsewhere, and the same problem exists?

**Review by yiguolei — APPROVED** [review]
LGTM

**Pull request merged**